code/modules/clothing/spacesuits/rig/modules/specific/ai_container.dm

```diff
@@ -175,41 +184,214 @@
+		to_chat(usr, SPAN_WARNING("You are not loaded into a hardsuit."))
+		return
+
+	var/obj/item/rig_module/ai_container/advanced/module = usr.loc.loc
+	if(!module.holder)
+		to_chat(usr, SPAN_WARNING("Your module is not installed in a hardsuit."))
+		return
+
+	module.jack_brain()
+
+/obj/item/rig_module/ai_container/advanced
+	name = "\improper AINI module"
+	desc = "An artificial intelligence neural integration module, suitable for most hardsuits."
+	description_fluff = "The unassuming device has no exterior markings, save for a network usage monitor and tiny warning label. On closer inspection, it reads, \"Warning: Device may cause damage if used without proper neural interface firmware.\" Something in your core urges you to heed that warning."
+	description_info = "Unless used by a robotic RIG pilot, or an organic one with a neural implant, the module will cause brain-damage on activation of its override mode."
+	icon_state = "AIIS"
+	var/controlling = FALSE
+
+	verb_holder = /obj/item/ai_verbs/advanced
+
+	var/last_defib = 1
+	var/mob/living/captive_brain/pilot_brain	
+	var/obj/item/weapon/shockpaddles/rig/paddles
+	var/imperfect_connection = FALSE	// Is the AI properly connected to the body? If N, they can't speak without sign-lang.
+	var/corpse_pilot = FALSE	// Is the AI controlling a corpse? If Y, kill the body again when they revert.
+
+/obj/item/rig_module/ai_container/advanced/Initialize()
+	. = ..()
+
+	paddles = new(src)
+
+/obj/item/rig_module/ai_container/advanced/removed()
+	revert()
+	..()
+
+/obj/item/rig_module/ai_container/advanced/process()
+	if(holder)
+		if((holder.offline || !holder.ai_override_enabled) && controlling)
+			revert()
+
+		if(controlling)
+			var/mob/living/carbon/human/H = holder.wearer
+
+			var/mob/living/simple_mob/animal/borer/Borer = H.has_brain_worms()
+			if(Borer)	// As funny as it would be to see a borer, AI, and human swap minds like musical chairs, let's not.
+				revert()
+				to_chat(integrated_ai, SPAN_DANGER("<span class='danger'>The neural jack signals a warning abruptly, before rapidly retracting.</span>"))
+				to_chat(H, SPAN_WARNING("You are shoved into consciousness as though you have been dropped into freezing water."))
+				return
+
+			var/obj/item/organ/internal/brain/Brain = H.internal_organs_by_name[O_BRAIN]
+			if(Brain)
+				if(Brain.robotic < ORGAN_ASSISTED)
+					H.add_modifier(/datum/modifier/mute, 30 SECONDS, suppress_failure = TRUE)
+
+			if(H.stat == DEAD && world.time > last_defib + 1 MINUTE && (H.getBruteLoss() + H.getFireLoss()) < (H.getMaxHealth() * 2 - H.getCloneLoss()))
+				H.setToxLoss(0)
+				paddles.attack(holder,integrated_ai)
+				if(H.stat != DEAD)
+					corpse_pilot = TRUE
+
+			if(H.stat != DEAD)
+				if(corpse_pilot)
+					H.setOxyLoss(0)
+					H.add_modifier(/datum/modifier/risen_corpse, 30 SECONDS, suppress_failure = TRUE)
+					last_defib = world.time
+				if(H.sleeping)
+					H.AdjustSleeping(-10)
```

Comment:
```suggestion
				if(H.stat != DEAD)
					H.setOxyLoss(0)
					H.add_modifier(/datum/modifier/risen_corpse, 30 SECONDS, suppress_failure = TRUE)
				else // They're still dead, defib failed: no point trying this again
					controlling = FALSE
			if(H.stat != DEAD && H.sleeping)
				H.adjustSleeping(-10)
```

Reply:
The defib does not occur until after the puppeteer is in the body, and corpse_pilot is a necessary var to ensure that a risen corpse properly _returns_ to corpsehood after the puppeteer lets go of the strings.

Reply:
So the risen corpse modifier only lasts for 30s, but the AI can control the mob for longer than that? This feels like a _really_ clunky way of handling whether the mob should fall back over or not, using both a modifier _and_ a variable on this module.

What would happen if the module or the rig were removed from the wearer (by any mechanism) while this is active?